shadowsocks-libev/files/shadowsocks.init

```diff
@@ -1,298 +1,512 @@
-		# Setup iptables rules
-		$IPTN -N SOCKS 2>/dev/null
-		ipset -! create shadowsocks_localnetwork hash:net
-		$IPTN -D SOCKS -m set --match-set shadowsocks_localnetwork dst -j RETURN 2>/dev/null
-		$IPTN -A SOCKS -m set --match-set shadowsocks_localnetwork dst -j RETURN
-		# Setup redirect rules with QoS
-		$IPTN -D SOCKS -p tcp -m dscp --dscp-class cs7 -j REDIRECT --to-ports $(( $lport + 1 )) 2>/dev/null
-		$IPTN -A SOCKS -p tcp -m dscp --dscp-class cs7 -j REDIRECT --to-ports $(( $lport + 1 ))
-
-		$IPTN -D SOCKS -p tcp -m dscp --dscp-class cs6 -j REDIRECT --to-ports $(( $lport + 2 )) 2>/dev/null
-		$IPTN -A SOCKS -p tcp -m dscp --dscp-class cs6 -j REDIRECT --to-ports $(( $lport + 2 ))
-
-		$IPTN -D SOCKS -p tcp -m dscp --dscp-class cs5 -j REDIRECT --to-ports $(( $lport + 3 )) 2>/dev/null
-		$IPTN -A SOCKS -p tcp -m dscp --dscp-class cs5 -j REDIRECT --to-ports $(( $lport + 3 ))
-
-		$IPTN -D SOCKS -p tcp -m dscp --dscp-class cs4 -j REDIRECT --to-ports $(( $lport + 4 )) 2>/dev/null
-		$IPTN -A SOCKS -p tcp -m dscp --dscp-class cs4 -j REDIRECT --to-ports $(( $lport + 4 ))
-
-		$IPTN -D SOCKS -p tcp -m dscp --dscp-class cs3 -j REDIRECT --to-ports $(( $lport + 5 )) 2>/dev/null
-		$IPTN -A SOCKS -p tcp -m dscp --dscp-class cs3 -j REDIRECT --to-ports $(( $lport + 5 ))
-
-		$IPTN -D SOCKS -p tcp -m dscp --dscp-class cs2 -j REDIRECT --to-ports $(( $lport + 6 )) 2>/dev/null
-		$IPTN -A SOCKS -p tcp -m dscp --dscp-class cs2 -j REDIRECT --to-ports $(( $lport + 6 ))
-
-		$IPTN -D SOCKS -p tcp -m dscp --dscp-class cs1 -j REDIRECT --to-ports $(( $lport + 7 )) 2>/dev/null
-		$IPTN -A SOCKS -p tcp -m dscp --dscp-class cs1 -j REDIRECT --to-ports $(( $lport + 7 ))
-
-		$IPTN -D SOCKS -p tcp -j REDIRECT --to-ports $lport 2>/dev/null
-		$IPTN -A SOCKS -p tcp -j REDIRECT --to-ports $lport
-	fi
-
-	if [ -n "$monitoring_ip" -a -x "$TRACK" ]; then
-		if ! $IPTN -C OUTPUT -d $monitoring_ip/32 -p tcp -m comment --comment "shadowsocks_tracker" -j REDIRECT --to-ports $(( $lport + 2 )) 2>/dev/null; then
-			$IPTN -A OUTPUT -d $monitoring_ip/32 -p tcp -m comment --comment "shadowsocks_tracker" -j REDIRECT --to-ports $(( $lport + 2 ))
-		fi
-		procd_open_instance
-		procd_set_param command "$TRACK" -r $reliability -c $count -t $tracktimeout -v $interval -o $down -u $up $monitoring_ip
-		procd_set_param file "$conf"
-		procd_set_param respawn
-		procd_set_param stderr 1
-		procd_close_instance
-	else
-		shadowsocks_enable_interfaces_hook
-	fi
+
+# This function launches one shadowsocks per CPU
+# $1 is a ref to the shadowsocks uci section (type =client)
+_launch_shadow() {
+    local conf cpu_count
+
+    # Determine the number of CPU cores
+    local cpu_sysfs="/sys/devices/system/cpu/"
+
+    if [ -d $cpu_sysfs ]; then
+        # Try to get the number of CPU cores
+        # We need to add a backslash before the + because busybox's find uses "posix-basic" regextype
+        cpu_count=$(find $cpu_sysfs -maxdepth 1 -type d -regex '.*/cpu[0-9]\+' | wc -l)
```

Comment:
This count dies cores and threads cores, not only real cores : You must count uniq cores id per physicals id.